Feature/local node message optimization

cryptowizard0 opened the pull request:
Description

feat: Optimization - Local node messages are now processed directly via channels instead of HTTP protocol  

• In outbox trySend, check if the target is the local node. If so, directly invoke node.Handle.  

• Subscription handling for outbox assignment results: A map (msgid -> chan) is added to the node. If outbox subscribes to assignment results, they are sent directly to outbox via the channel. (map[string]chan schema.AssignmentResult)  

• For local messages, outbox waits for the assignment channel result after calling node.Handle.  

• Retry mechanism: Timeout or failure triggers up to 5 retry attempts.

cryptowizard0 committed:
add local message sending

cryptowizard0 committed:
Update outbox.go

cryptowizard0 commented:
HIP Link:
[HIP 2](https://github.com/hymatrix/hymx/discussions/7)

outprog commented on node/outbox.go:
```diff
@@ -100,19 +100,95 @@ func (n *Node) trySend(pid, target string) {
 	}
 }
 
+func (n *Node) tryGetLocalAssign(msgId string, nodes []registrySchema.Node, item goarSchema.BundleItem) (assignItem goarSchema.BundleItem, err error) {
```
It seems unnecessary to pass the nodes parameter when sending to self. Please handle the self-send case separately, so that nodes is either not required or explicitly ignored in this scenario.

outprog commented on node/outbox.go:
```diff
@@ -100,19 +100,95 @@ func (n *Node) trySend(pid, target string) {
 
+func (n *Node) tryGetLocalAssign(msgId string, nodes []registrySchema.Node, item goarSchema.BundleItem) (assignItem goarSchema.BundleItem, err error) {
+	// make channel for assignment result
+	resultChan := make(chan schema.AssignmentResult, 1)
```
The channel buffer size of 1 may lead to performance bottlenecks under heavy load. Consider increasing it (e.g., 1000+) to reduce the risk of blocking when processing results concurrently.

cryptowizard0 committed:
refactor: implement callback-based assignment result handling

Refactored assignment result handling from channel-based to callback-based approach for better resource management and performance.

- Restored assignmentChan structure in node.go
- Added RemoveAssignmentHandler function for proper cleanup
- Modified tryGetLocalAssign to use temporary assignment handlers
- Removed outboxAssignmentChans and related code
- Updated comments to English for better code consistency
- Ensured proper handler registration and cleanup in outbox.go

This change improves memory management by preventing handler leaks and simplifies the assignment result processing architecture.

outprog commented on node/assignment.go:
```diff
@@ -54,3 +63,21 @@ func (n *Node) signAssign(pid, msgid string, nonce int64) (assign hymxSchema.Ass
+	select {
+	case n.assignmentChan <- assignmentResult:
+		// send success to channel
+	default:
```
should remove

outprog commented on node/assignment.go:
```diff
@@ -54,3 +63,21 @@ func (n *Node) signAssign(pid, msgid string, nonce int64) (assign hymxSchema.Ass
+		AssignItem: assignItem,
+		Error:      err,
+		Pid:        pid,
+		ItemId:     itemId,
```
use item.

cryptowizard0 committed:
refactor：refactor AssignmentResult structure and optimize related processing logic

- Move ItemId field to Item struct and adjust AssignmentResult struct field order
- Optimize channel processing logic in tryGetLocalAssign method to avoid potential blocking
- Simplify sendAssignmentResult method by removing unnecessary select statements
- Update related logs and variable references to match new struct definitions

cryptowizard0 committed:
refactor: code struct optimize

- Move sendAssignmentResult out of assignment
- Remove dup params

cryptowizard0 committed:
refactor：optimize chan

cryptowizard0 committed:
refactor: remove sendAssignmentResult function
